keps/sig-network/0752-endpointslices/README.md

```diff
@@ -304,6 +301,94 @@ following well known key:
 If the Kubernetes Service has topological keys specified, the corresponding node
 labels will be copied to endpoint topology.
 
+### Converting Topology in EndpointSlice GA
+
+For beta resources that have set the topology field, the string to string map
+will be converted into an annotation on the v1 resource.
+
+The size limits of the topology field is 63B per key and per value, with a
+maximum of 126B per key-value pair. The maximum number of endpoints in an
+EndpointSlice is 1000.
+
+```
+{Max Key-Value Pairs} * ({Max Key Size} + {Max Value Size}) * {Max Endpoints}
+16 * (63 + 63) * 1000 = 2,016,000b
+```
+
+The maximum value of the combined maps in the endpoints is almost a factor of
+10 greater than the max annotation size (256kB). To reduce the size of the
+annotation, only the first 100 endpoints will be preserved per EndpointSlice
+and each endpoint's map will be base64 encoded.
+
+The annotation key will be **topology.kubernetes.io** and the value will be a
+json string where the keys will be the index of the Endpoint in the slice. The
+value will have the following format:
+```
+{"0":b64EncodedMap(0th Endpoint),...,"n":b64EncodedMap(nth Endpoint)}
+```
+where `n < 100`.
+
+The map will be encoded by the following:
+```
+import (
+    "encoding/base64"
+    "encoding/json"
+)
+
+func encodeMap(topology map[string]string) (string, error) {
+    mapBytes, err := json.Marshal(topology)
+    if err != nil {
+        return "", err
+    }
+
+    return base64.StdEncoding.EncodeToSTring(mapBytes), nil
+}
+```
+and consumers will be able to decode with the following:
+```
+import (
+    "encoding/base64"
+    "encoding/json"
+)
+
+func decodeMap(encodedMap string) (map[string]string, error) {
+    decodedBytes, err := base64.StdEncoding.DecodeString(encodedMap)
+    if err != nil {
+        return nil, err
+    }
+
+    decodedMap := make(map[string]string)
+    err := json.Unmarshal(decodedBytes, &decodedMap)
+    return decodedMap, err
+}
+```
+
+For example if the 0th Endpoint had the following map:
+```
+{
+  "key-1": "value-1",
+  "key-2": "value-2",
+  "key-3": "value-3",
+}
+```
+The annotation would look like:
+
+```
+topology.kubernetes.io:`{"0":"eyJrZXktMSI6InZhbHVlLTEiLCJrZXktMiI6InZhbHVlLTIiLCJrZXktMyI6InZhbHVlLTMifQ=="}`
+```
+
+The only key that will not be preserved in the annotation will be
+**topology.kubernetes.io/zone** and instead its value will be added on the
+[zone field](#zone-per-endpoint) on the Endpoint.
+
+
```

Comment:
It may be worth adding a note somewhere that this could potentially get out of sync as endpoints are rearranged in an EndpointSlice.

Reply:
The annotation has to be updated if endpoints are re-arranged

Reply:
This feels like a very significant tradeoff of dropping the topology field. Will consumers of the v1 API always have to include some special logic for these endpoints? Do we need to prevent this annotation from being set with the v1beta1 API?

Reply:
We would need a better identifier for the endpoint key. I am not sure how we could update the annotation if the order changes on a v1 version. On a v1beta1 resource, it is not an issue since we could just reconstruct the annotation.

Reply:
v1 clients in a mixed mode need something like an unpack function and a mapping per endpoint -> annotation payload.  If they edit a slice, they need to re-assemble the annotation.  That doesn't sound so bad?

Reply:
That approach seems reasonable and we just need to make sure it is well documented.